src/tree/tree_model.cc

```diff
@@ -777,8 +862,17 @@ void RegTree::SaveModel(Json* p_out) const {
   std::vector<Json> indices(n_nodes);
   std::vector<Json> conds(n_nodes);
   std::vector<Json> default_left(n_nodes);
+  std::vector<Json> split_type(n_nodes);
 
-  for (int32_t i = 0; i < n_nodes; ++i) {
+  std::vector<Json> cat_seg_begin(n_nodes);
+  std::vector<Json> cat_seg_size(n_nodes);
+  std::vector<Json> categories(n_nodes);
+
+  auto& self = *this;
```

Comment:
> Why not use the customary this pointer directly? 

I think at some point I use `operator[]` then removed the usage.  The model IO is revised in https://github.com/dmlc/xgboost/pull/6028 and stores actual categories instead of bitset.  I will address related comments in there.  Rebasing this PR becomes quite difficult.